### PR TITLE
Adds option to limit sling unhatched duration

### PR DESCRIPTION
**Pull request by Kyep**
🆑 Kyep
add: Added config option to limit the amount of time shadowlings can remain unhatched before they start getting pushed to hatch.
/🆑

Currently: 
- Shadowlings can remain unhatched indefinitely.
- There is no way to detect an unhatched shadowling.
- There is usually one unhatched sling who sticks around, not doing anything, long after the rest have died, artificially prolonging the round for everyone.
- This makes shadowling rounds dull. It would be better if that last shadowling was pushed to hatch, and maybe do something, rather than wasting time for everyone.

With this PR:
- A new config option, SHADOWLING_MAX_AGE is added.
- In shadowling rounds, shadowlings who are alive at least that long without hatching start getting messages prompting them to hatch, as well as occasional minor burn damage.
- The default setting for this is 1800 (aka: 30 minutes) but you can set it to whatever you like. Server config option.

**Code logic review appreciated.**

### CODE REVIEW

**Comment by Spartan6** [comment]
Praise be.

**Comment by FlattestGuitar** [comment]
I'd add some sort of a non-ambiguous hint for people so that they don't randomly hatch in the middle of a hallway.

Also, this makes mid-round slings being added impossible, since the timer does not take the slings' creation time into account; just general world time. Would be good to only impact roundstart slings or add a timer to the shadowling datum itself.

**Comment by Kyep** [comment]
@FlattestGuitar I thought about that... but the damage is currently so minor (1 point, every second or two) that so long as you hatch quickly after being made a sling, it isn't going to hurt you. Especially since you're +GODMODE for a fair part of the time you're actually hatching. If you hatch quickly, you'll take less than 20 points of damage before you hatch, which will be VERY quickly healed during your time as a sling. 

So yeah, could add a per-sling timer, but given that:
- they should be hatching anyway, and assuming they do, they won't suffer any ill effects from a few points of brute that sling status will rapidly heal
- you control where they spawn, so you can easily spawn them in a perfect-for-hatching location... its not like they should have to move somewhere to hatch

File: code/game/gamemodes/shadowling/shadowling.dm [hunk 1]
```diff
@@ -205,6 +205,13 @@ Made by Xhuis
 			continue
 		if(shadow.current.stat == DEAD)
 			continue
+		if(world.time - round_start_time > (config.shadowling_max_age * 10))
+			if(ishuman(shadow.current))
+				var/mob/living/carbon/human/H = shadow.current
+				if(!istype(H.species, /datum/species/shadow))
+					H.take_overall_damage(1, 0)
+					to_chat(H, "<span class='userdanger'>You cannot contain your shadowling essense any longer! You must hatch, or die!</span>")
```
Comment (Spartan6) [inline code/game/gamemodes/shadowling/shadowling.dm]: ‘You can barely hold yourself in this lesser form!’
‘The urge to become something greater is overwhelming!’
‘You feel a passion to hatch free of this shell to assume godhood!’

Comment (Kyep) [reply code/game/gamemodes/shadowling/shadowling.dm]: Yeah, that's probably a good idea. Added.

**Comment by Denthamos** [comment]
Thank goodness.  No more unhatched slings waiting till the 13:30 mark just to hatch and try to rush it or just prolonging the round unnecessarily.

**Comment by Kyep** [comment]
Now, the timer is now based on a 'cycles_unused' var of the shadowling mind's shadowling_hatch ability.

Implications:
- It is absolutely impossible to get this message/damage unless you have the sling hatch ability. Rehatch does not count. Neither does being a thrall. Nor does merely having the sling species. You have to actually be able to hatch in order to have your counter incremented, or, if it is already too big, trigger the warning.
- The timer is based on the amount of time since you gained the sling hatch ability. So for example roundstart slings have 30m (by default) from round start, but later-added slings have 30m from when they become slings.

**Comment by imsxz** [comment]
There is actually a way to detect unhatched shadowlings. Try to mindshield them, and they'll resist it, the same way revheads resist them.

**Comment by taukausanake** [comment]
@imsxz I think that falls into the realm of metagaming even in this context. Mindshielding someone to find out if they are an antag isn't good

**Comment by KingPhilipIII** [comment]
This deserves a balance tag imo

**Comment by tigercat2000** [comment]
As it is, this affects balance by changing the normal shadowling experience - @Kyep Could you please add a special value that disables this behavior, such as -1, and make that the default? We can experiment with configuration changes on the live server.

**Comment by Kyep** [comment]
@tigercat2000 the default value (with no directive defined in config) is now zero.
With that left as-is, this check never passes:
"if(shadow.special_role == SPECIAL_ROLE_SHADOWLING && config.shadowling_max_age)"
and the code never runs.

File: config/example/config.txt [hunk 2]
```diff
@@ -95,6 +95,9 @@ PROBABILITY EXTENDED 2
 PROBABILITY NUCLEAR 3
 PROBABILITY ABDUCTION 0
 
+# MAXIMUM SECONDS SHADOWLINGS CAN REMAIN UNHATCHED BEFORE THEY TAKE DAMAGE
+SHADOWLING_MAX_AGE 1800
```
Comment (tigercat2000) [inline config/example/config.txt]: Change this default to 0 as well, please.

**Comment by Kyep** [comment]
@tigercat2000 Changed

File: code/game/gamemodes/shadowling/shadowling.dm [hunk 3]
```diff
@@ -201,11 +201,23 @@ Made by Xhuis
...
-		if(!istype(shadow.current,/mob/living/carbon/human) && !istype(shadow.current,/mob/living/simple_animal/ascendant_shadowling))
+		if(!ishuman(shadow.current) && !istype(shadow.current,/mob/living/simple_animal/ascendant_shadowling))
 			continue
 		if(shadow.current.stat == DEAD)
 			continue
 		shadows_alive++
+		if(shadow.special_role == SPECIAL_ROLE_SHADOWLING && config.shadowling_max_age)
+			if(ishuman(shadow.current))
+				var/mob/living/carbon/human/H = shadow.current
+				if(!istype(H.species, /datum/species/shadow))
```
Comment (tigercat2000) [inline code/game/gamemodes/shadowling/shadowling.dm]: This should be `if(H.get_species() != "Shadow")` (I realize magic strings are usually worse than an istype check, but this is the consistent way of doing it through the rest of the code, and we'll probably need to overhaul get_species at some point)

**Comment by Kyep** [comment]
@tigercat2000 Changed

**Pull request merged**